### ISSUE TITLE
Set up gerrit integration test / demo

### ISSUE THREAD

**Issue by haya14busa** [reviewdog/reviewdog#581]
Maybe we can use  http://gerrithub.io/.

**Comment by felipecassiors** [comment]
Gerrit has good integration with Jenkins and [Zuul](https://zuul-ci.org/), but so far I never saw a free hosted CI tool that supports Gerrit.

**Comment by haya14busa** [comment]
We don't have Jenkins instance unfortunetely.

**Comment by haya14busa** [comment]
I think it's ideal If we can sync the code / Pull Request from GitHub Actions, but I'm not still sure whether it's possible with gerrithub.

**Comment by felipecassiors** [comment]
@haya14busa sync pull requests with GerritHub is possible, however, that doesn't solve the problem... The CI must be triggered by Gerrit, since it must know the change number and patch set number. I'll take a look later today in the feasibility of some kind of GitHub Actions integration with Gerrit.

**Comment by felipecassiors** [comment]
I studied three different solutions.

1. Gerrit triggering a GitHub Actions workflow through webhook
    - Requires a webhook interceptor for transformation, since the outcoming webhook of Gerrit can't be used as an input for GitHub Repository Dispatch API. I even [wrote an LUA function](https://gist.github.com/felipecassiors/d17c7c17ce45ae5989b2d23e1d78602f) to use in [Webhook Relay](https://webhookrelay.com/), which works, but this solution needs an external (and not free) service.

2. A new GitHub Actions workflow triggered on every push or pull request that polls GerritHub to check if there is a Gerrit Change for that commit or not (because GerritHub replicates back every commit to GitHub).

3. I don't exactly what do we want to achieve, but if is automatic tests, I thought about something like the following, in pull requests: 
    1. Deploy a temporary test instance somewhere of Gerrit with the current base branch state of the repository 
    2. Automatically create a change 
    3. Run `reviewdog` from the same pipeline, pointing to the temporary Gerrit instance
    But what do we want to achieve? Visual feedback on Gerrit? Or just an automatic test expecting `reviewdog` to return an error?

**Comment by haya14busa** [comment]
Thank you @felipecassiors for the investigation! It's very helpful.

1. Interestingly, I wrote a similar Google Apps script which intercepts GitHub Webhook and sends Dispatch API yesterday (https://github.com/reviewdog/inviter/blob/master/gh-apps-script-proxy/Code.js).
We can use Google Apps Script for free, so maybe we can adopt this option.

2. Alternatively, we may want to use label/issue_comment/pull_request_comment events to trigger the workflow manually for gerrit. It's not ideal, but it's better than nothing and in fact, we don't need to test gerrit in most cases too.

3. Do you think it's easy to deploy a temporary test instance of Gerrit?

I'd like to archive some visual feedback on Gerrit. i.e. I'd like to see comments posted from reviewdog.
Ideally, it's better to check it automatically by fetching comments from API after posting comments for example, but we didn't do this for other reporters like GitHub and GitLab, so it's optional.


In summary, I think we can adopt option 1 with Apps script, what do you think?

**Comment by felipecassiors** [comment]
In this case, for numbers 1 and 2 there are still a few open points.

- I'm not sure if we can have GitHub and GerritHub working together. Their documentation says to use GerritHub for pushing and merging code, and never directly on GitHub anymore. [See](https://review.gerrithub.io/static/intro.html). If this happens, you need to manually run a few `git` commands to merge the code back on GerritHub. **If that is true, it will become a little inadequate to use GerritHub for visualization, will further check to see.**

For number 3:

- Yes, via docker: https://hub.docker.com/r/gerritcodereview/gerrit. And it's totally possible to use Gerrit's API to fetch the comments and perform an automatic validation. See https://gerrit-review.googlesource.com/Documentation/rest-api-changes.html#list-comments.

**Comment by haya14busa** [comment]
hmm... we basically use pull-request & merge for reviewdog development instead of pushing commit to master directlry, so if it works well with gerrithub, then it should be fine.
Or we can create a different repo for demo/testing purpose. Maybe we should check it with a different repo first at least.

**Comment by felipecassiors** [comment]
I just tested here, no matter if the change is directly pushed to the master branch or via a pull request. If the merge happens in GitHub and not in GerritHub, the repository in GerritHub doesn't get updated, and so forth, you need to manually sync it again.

> How can I resume replication to GitHub ?
> If you pushed directly on GitHub, in order to allow Gerrit catching-up with the new commits on > GitHub follow the steps below:
> 
> Clone the repository from Gerrit (git clone <GerritHub URL> master)
> Pull and merge the latest GitHub changes (git pull <GitHub URL> master)
> Push the merged branch to Gerrit (git push <GerritHub URL> master)
From: https://review.gerrithub.io/static/intro.html

Seems like GerritHub is only feasible for people who want to mainly use Gerrit for Code Review. It can import GitHub Pull Requests as Changes in Gerrit, but still, you have to merge them through Gerrit.

Unfortunately. I'll send an email for them to check if there is any plan to support this reverse sync. However, the different repo (and always use Gerrit for it) should work.

**Comment by felipecassiors** [comment]
By any chance, @haya14busa, do you know any free service who would let us run a temporary server via a docker container for testing pull requests?

**Comment by haya14busa** [comment]
Unfortunately no.
I think we can start by creating a different repository.

**Comment by felipecassiors** [comment]
Yes, I also think so.
